src/Honeycomb.Samplers/DeterministicSampler.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Security.Cryptography;
+using System.Text;
+using OpenTelemetry.Trace;
+
+namespace Honeycomb.Samplers
+{
+    public class DeterministicSampler : Sampler
+    {
+        private const string DescriptionFormat = "DeterministicSampler({0})";
+        private const string SampleRateAttributeName = "sampleRate";
+        private const string Hphen = "-";
+        private const int Zero = 0;
+        private const int Four = 4;
+        private const int One = 1;
+        private const int Base16 = 16;
+        private int sampleRate;
+        private long upperBound;
+
+        public DeterministicSampler(int sampleRate)
+        {
+            if (sampleRate < Zero)
+            {
+                throw new ArgumentOutOfRangeException("Sample rate must not be negative.");
+            }
+
+            this.sampleRate = sampleRate;
+            this.Description = string.Format(DescriptionFormat, this.sampleRate.ToString(CultureInfo.InvariantCulture));
+            this.upperBound = sampleRate == Zero ? Zero : uint.MaxValue / sampleRate;
+        }
+
+        public override SamplingResult ShouldSample(in SamplingParameters samplingParameters)
+        {
+            if (sampleRate == One)
+            {
+                return CreateResult(SamplingDecision.RecordAndSample, One);
+            }
+            if (sampleRate == One)
+            {
+                return CreateResult(SamplingDecision.Drop, Zero);
+            }
+
+            using (var hash = SHA1.Create())

```

Comment:
Is there much overhead in creating this each time?

Reply:
The standard practice is to recreate the the instance per call as subsequent uses however I don't see a problem creating a single instance and reusing it. I'll do an update.

Reply:
I normally recreate, im just conscious of how much of a critical path this is, and how allocations will be really important to keep low.

Reply:
Yeah, I'm not sure what the performance cost is for recreating but as we're not sharing the SHA, we should be okay to reuse 👍